data/lib/index.js

```diff
@@ -0,0 +1,139 @@
+import { pipe } from 'it-pipe'
+import { batch } from 'streaming-iterables'
+import { CID } from 'multiformats/cid'
+
+import { MAX_BATCH_GET_ITEMS } from '../tables/constants.js'
+import { createCarTable } from '../tables/car.js'
+import { createFerryTable } from '../tables/ferry.js'
+import { createAggregateService } from './aggregate-service.js'
+
+/**
+ * @typedef {import('../types.js').FerryOpts} FerryOpts
+ * @typedef {import('../types.js').CarItem} CarItem
+ * @typedef {import('../types.js').CarItemFerry} CarItemFerry
+ *
+ * @typedef {object} FerryCtx
+ * @property {string} region
+ * @property {string} tableName
+ * @property {FerryOpts} [options]
+ * 
+ * @typedef {object} CarTableCtx
+ * @property {string} region
+ * @property {string} tableName
+ * @property {import('../types.js').CarOpts} [options]
+ */
+
+/**
+ * Add cars to a loading ferry.
+ *
+ * @param {CarItemFerry[]} cars
+ * @param {FerryCtx} ferryCtx
+ */
+export async function addCarsToFerry (cars, ferryCtx) {
+  const ferryTable = createFerryTable(ferryCtx.region, ferryCtx.tableName, ferryCtx.options)
+
+  const ferryId = await ferryTable.getFerryLoading()
+  await ferryTable.addCargo(ferryId, cars)
+
+  return {
+    id: ferryId
+  }
+}
+
+/**
+ * Sets current Ferry as ready if not previously done
+ *
+ * @param {string} ferryId
+ * @param {FerryCtx} ferryCtx
+ */
+export async function setFerryAsReady (ferryId, ferryCtx) {
+  const ferryTable = createFerryTable(ferryCtx.region, ferryCtx.tableName, ferryCtx.options)
+
+  // Update state of ferry to ready
+  try {
+    await ferryTable.setAsReady(ferryId)
+  } catch (/** @type {any} */ error) {
+    // If error is for condition we can safely ignore it given this was changed in a concurrent operation
+    if (error.name !== 'ConditionalCheckFailedException') {
+      throw error
+    }
+  }
+}
+
+/**
+ * Sets current Ferry offer.
+ *
+ * @param {string} ferryId
+ * @param {object} ctx
+ * @param {FerryCtx} ctx.car
+ * @param {FerryCtx} ctx.ferry
+ * @param {import('../types.js').StorefrontSignerCtx} ctx.storefront
+ * @param {import('@ucanto/principal/ed25519').ConnectionView<any>} ctx.aggregationServiceConnection
+ */
+export async function setFerryOffer (ferryId, ctx) {
+  const carTable = createCarTable(ctx.car.region, ctx.car.tableName, ctx.car.options)
+  const ferryTable = createFerryTable(ctx.ferry.region, ctx.ferry.tableName, ctx.ferry.options)
+  const aggregateService = await createAggregateService(ctx.storefront, ctx.aggregationServiceConnection)
+
+  // Create Offer
+  /** @type {CarItem[]} */
+  const offers = await pipe(
```

Comment:
Needed to get CAR details of all the cargo in ferry to include in `aggregate/offer`

Reply:
I will add the comment for this. There are comments in `data-stack` with information about role of each table in the picture and architecture diagram that was previously shared and presented in Demo, if you would like more context https://www.notion.so/w3filecoin-pipeline-architecture-204bb37b550b4e46ab9ea0799396d96b . This document can be moved into Github as well.